In [2]:
pip install pandasql

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [23]:
from pandasql import sqldf
import sqlalchemy
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns', 50)

# **Data Exploration**

In [4]:
horses = pd.read_csv('Horses.csv')

In [20]:
horses.head()

RaceID  HorseID  Cloth  Stall  WeightValue   JockeyID  AllowanceValue  \
0  989418  2963489      7    5.0          132  1398052.0             NaN   
1  989418  2972665     13    6.0          132  1165412.0             NaN   
2  989418  2972662      9    7.0          132  1239852.0             NaN   
3  989418  2960365      1   12.0          132  1417524.0             NaN   
4  989418  2968596     11    8.0          132  1354773.0             NaN   

   OverweightValue  TrainerID    OwnerID  LastRunDaysFlat  Age  Blinkers  \
0              NaN    1198534  1289143.0             26.0  3.0       NaN   
1              NaN    1188128  1140080.0              NaN  3.0       NaN   
2              NaN    1199122  1448576.0              NaN  3.0       NaN   
3              NaN    1049637  1487690.0             33.0  3.0       NaN   
4              NaN    1140041  1403676.0             24.0  3.0       NaN   

   Hood  Visor  EyeShield  Eyecover  TongueStrap  CheekPieces ForecastPrice  \
0   NaN    NaN        NaN       NaN          NaN          NaN        16-Jan   
1   NaN    NaN        NaN       NaN          NaN          NaN        20-Jan   
2   NaN    NaN        NaN       NaN          NaN          NaN        10-Jan   
3   NaN    NaN        NaN       NaN          NaN          NaN        07-Apr   
4   NaN    NaN        NaN       NaN          NaN          NaN        08-Jan   

   StatCourse  StatDistance  StatCourseDistance StartingPrice Bred      DamID  \
0         NaN           NaN                 NaN        12-Jan  SAF  2803520.0   
1         NaN           NaN                 NaN          40/1  SAF  2749301.0   
2         NaN           NaN                 NaN           NaN  SAF  2761722.0   
3         NaN           NaN                 NaN        11-Oct  SAF  2347111.0   
4         NaN           NaN                 NaN           NaN  SAF  2618721.0   

      SireID  DamSireID Sex Colour  YearBorn  MeetingID  RaceNumber  \
0  2004880.0  2179351.0   f      b       NaN      87048           1   
1  2960049.0  2001540.0   f      b       NaN      87048           1   
2  2441786.0  2123459.0   f      b       NaN      87048           1   
3  2170013.0  2071137.0   f      b       NaN      87048           1   
4  2410998.0  2170115.0   f      b       NaN      87048           1   

  ScheduledTime  HorseCount  Runners Going Weather  Course_Distance RaceType  \
0    ##########          13     11.0  Good  Cloudy             1312     Flat   
1    ##########          13     11.0  Good  Cloudy             1312     Flat   
2    ##########          13     11.0  Good  Cloudy             1312     Flat   
3    ##########          13     11.0  Good  Cloudy             1312     Flat   
4    ##########          13     11.0  Good  Cloudy             1312     Flat   

  TrackType  Handicap eligibility  CourseID MeetingDate  Won  
0      Turf       NaN    2YO only       380  02/05/2017    0  
1      Turf       NaN    2YO only       380  02/05/2017    0  
2      Turf       NaN    2YO only       380  02/05/2017    0  
3      Turf       NaN    2YO only       380  02/05/2017    1  
4      Turf       NaN    2YO only       380  02/05/2017    0

In [5]:
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36819 entries, 0 to 36818
Data columns (total 46 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   RaceID              36819 non-null  int64  
 1   HorseID             36819 non-null  int64  
 2   Cloth               36819 non-null  int64  
 3   Stall               36615 non-null  float64
 4   WeightValue         36819 non-null  int64  
 5   JockeyID            36818 non-null  float64
 6   AllowanceValue      1609 non-null   float64
 7   OverweightValue     106 non-null    float64
 8   TrainerID           36819 non-null  int64  
 9   OwnerID             36616 non-null  float64
 10  LastRunDaysFlat     34300 non-null  float64
 11  Age                 36616 non-null  float64
 12  Blinkers            10342 non-null  float64
 13  Hood                0 non-null      float64
 14  Visor               68 non-null     float64
 15  EyeShield           0 non-null      float64
 16  Eyec

Here, I can see at a glance, that we have 5 columns that have all null values in them. I also notice many other columns also contain lots of null values, as they have a count less than 36819 non-null values.
Questions raised:
- What does AllowanceValue mean and is this required?
- Is OverweightValue required - lots of null values?
- YearBorn - this has lots of null values whereas "Age" doesn't
- StatCourse/StatDistance/StatCourseDistance - all 3 have lots of null values - do we need these?
- Should "cloth" be numerical? If so, is there a mapping to a categorical description?

In [7]:
SELECT COUNT(DISTINCT RaceID) as "Number of Races" FROM horses;

Number of Races
0             3000

In [9]:
SELECT COUNT(DISTINCT HorseID) as "Number of different Horses" FROM horses;

Number of Horses
0              6064

In [17]:
SELECT Count(DISTINCT Bred)
FROM horses

count(DISTINCT Bred)
0                    13

In [19]:
SELECT distinct Sex
FROM horses

Sex
0     f
1     c
2     m
3  None
4     g
5     h

In [16]:
SELECT distinct Colour
FROM horses

Colour
0      b
1     gr
2     br
3    dkb
4    bl/
5   None
6     ch
7    b/b
8      g
9    gr/

## Statistics

- Number of races: 3000
- Number of different horses: 6064
- Number of different countries where bred: 13
- Number of different sexes: 5
- Number of different colours: 9

Questions/concerns raised:
- ForecastPrice and StartingPrice have datatypes "Object", which are showing as a date either in the form DD-Month or ##/#. These should be corrected to all be the same format and ensure it is accurate.
- Sex: should there be 5 different options for sex and what do these mean/can we categorise them more to just f/m?
Colour: what do these mean, can we rename to make it clearer for end user?
- ScheduledTime: Data appearing as ###, instead of a timestamp
- StatCourseDistance has lots of null values, there is also a column called "CourseDistance", are these related and how?

In [7]:
# Dropping null/inaccurate/non-relevant columns 
horses[['RaceID', 'HorseID', 'Cloth', 'Stall', 'WeightValue', 'JockeyID',
       'AllowanceValue', 'OverweightValue', 'TrainerID', 'OwnerID',
       'LastRunDaysFlat', 'Age', 
       # 'Blinkers', 'Hood', 'Visor', 'EyeShield', 'Eyecover', 'TongueStrap', 'CheekPieces','ForecastPrice', 
       # 'StatCourse', 'StatDistance', 'StatCourseDistance', 
        # 'StartingPrice', 
        'Bred', 'DamID', 'SireID', 'DamSireID', 'Sex', 'Colour', 
        # 'YearBorn', 
        'MeetingID', 'RaceNumber',
        # 'ScheduledTime', 
        'HorseCount', 'Runners', 'Going',
        'Weather', 'Course_Distance', 'RaceType', 'TrackType', 'Handicap',
        'eligibility', 'CourseID', 'MeetingDate', 'Won']]

RaceID  HorseID  Cloth  Stall  WeightValue   JockeyID  AllowanceValue  \
0       989418  2963489      7    5.0          132  1398052.0             NaN   
1       989418  2972665     13    6.0          132  1165412.0             NaN   
2       989418  2972662      9    7.0          132  1239852.0             NaN   
3       989418  2960365      1   12.0          132  1417524.0             NaN   
4       989418  2968596     11    8.0          132  1354773.0             NaN   
...        ...      ...    ...    ...          ...        ...             ...   
36814  1028128  2980319      8    5.0          132  1165412.0             NaN   
36815  1028128  2985033      6    3.0          132  1079281.0             NaN   
36816  1028128  2977403      7    1.0          132  1440495.0             NaN   
36817  1028128  2978297      4    4.0          132  1288634.0             NaN   
36818  1028128  3007603      9    9.0          132  1200677.0             NaN   

       OverweightValue  TrainerID    OwnerID  LastRunDaysFlat  Age Bred  \
0                  NaN    1198534  1289143.0             26.0  3.0  SAF   
1                  NaN    1188128  1140080.0              NaN  3.0  SAF   
2                  NaN    1199122  1448576.0              NaN  3.0  SAF   
3                  NaN    1049637  1487690.0             33.0  3.0  SAF   
4                  NaN    1140041  1403676.0             24.0  3.0  SAF   
...                ...        ...        ...              ...  ...  ...   
36814              NaN    1079200  1492946.0             45.0  4.0  SAF   
36815              NaN    1403297  1494400.0             28.0  4.0  SAF   
36816              NaN    1079200  1482459.0             28.0  4.0  SAF   
36817              NaN    1079234  1492370.0             24.0  4.0  SAF   
36818              NaN    1491252  1290304.0             32.0  4.0  SAF   

           DamID     SireID  DamSireID Sex Colour  MeetingID  RaceNumber  \
0      2803520.0  2004880.0  2179351.0   f      b      87048           1   
1      2749301.0  2960049.0  2001540.0   f      b      87048           1   
2      2761722.0  2441786.0  2123459.0   f      b      87048           1   
3      2347111.0  2170013.0  2071137.0   f      b      87048           1   
4      2618721.0  2410998.0  2170115.0   f      b      87048           1   
...          ...        ...        ...  ..    ...        ...         ...   
36814  2626929.0  2005847.0  2167985.0   g      b      91584           2   
36815  2632498.0  2058573.0  2000564.0   g      b      91584           2   
36816  2653726.0  2854936.0  2264809.0   g     ch      91584           2   
36817  2736480.0  2699207.0  2705498.0   g      b      91584           2   
36818  2618887.0  2345894.0  2129959.0   c      b      91584           2   

       HorseCount  Runners     Going Weather  Course_Distance RaceType  \
0              13     11.0      Good  Cloudy             1312     Flat   
1              13     11.0      Good  Cloudy             1312     Flat   
2              13     11.0      Good  Cloudy             1312     Flat   
3              13     11.0      Good  Cloudy             1312     Flat   
4              13     11.0      Good  Cloudy             1312     Flat   
...           ...      ...       ...     ...              ...      ...   
36814           9      8.0  Standard  Cloudy             1969     Flat   
36815           9      8.0  Standard  Cloudy             1969     Flat   
36816           9      8.0  Standard  Cloudy             1969     Flat   
36817           9      8.0  Standard  Cloudy             1969     Flat   
36818           9      8.0  Standard  Cloudy             1969     Flat   

       TrackType  Handicap eligibility  CourseID MeetingDate  Won  
0           Turf       NaN    2YO only       380  02/05/2017    0  
1           Turf       NaN    2YO only       380  02/05/2017    0  
2           Turf       NaN    2YO only       380  02/05/2017    0  
3           Turf       NaN    2YO only       380  02/05/2017

# Initial analysis

In [38]:
SELECT Age, 
COUNT(Won) as Number_of_won_races, 
count(*) * 100.0 / sum(count(*)) over() as win_percentage
FROM horses
WHERE won = 1
GROUP BY Age
ORDER BY Age

Age  Number_of_won_races  win_percentage
0  2.0                   14        0.481596
1  3.0                  936       32.198142
2  4.0                 1056       36.326109
3  5.0                  552       18.988648
4  6.0                  238        8.187135
5  7.0                   64        2.201582
6  8.0                   24        0.825593
7  9.0                    6        0.206398
8  NaN                   17        0.584795

In [39]:
SELECT WeightValue, 
COUNT(Won) as Number_of_won_races, 
count(*) * 100.0 / sum(count(*)) over() as win_percentage
FROM horses
WHERE won = 1
GROUP BY WeightValue
ORDER BY WeightValue

WeightValue  Number_of_won_races  win_percentage
0           106                    2        0.068799
1           107                    2        0.068799
2           109                    2        0.068799
3           110                    4        0.137599
4           111                    8        0.275198
5           112                    1        0.034400
6           114                    4        0.137599
7           115                   89        3.061576
8           116                   54        1.857585
9           117                   51        1.754386
10          118                   53        1.823185
11          119                   77        2.648779
12          120                   54        1.857585
13          121                   83        2.855177
14          122                   77        2.648779
15          123                  135        4.643963
16          125                   97        3.336773
17          126                  218        7.499140
18          127                  116        3.990368
19          128                  279        9.597523
20          129                  122        4.196766
21          130                  215        7.395941
22          131                   85        2.923977
23          132                  946       32.542140
24          133                   52        1.788786
25          135                   37        1.272790
26          136                   25        0.859993
27          137                   16        0.550396
28          138                    1        0.034400
29          139                    2        0.068799